data/src/main/java/org/apache/iceberg/data/DeleteFilter.java

```diff
@@ -185,14 +203,35 @@ private CloseableIterable<T> applyEqDeletes(CloseableIterable<T> records) {
         .reduce(Predicate::and)
         .orElse(t -> true);
 
-    Filter<T> remainingRowsFilter = new Filter<T>() {
+    Filter<T> remainingRowsFilter = hasMetadataColumnIsDeleted ? getMarker(remainingRows) : getFilter(remainingRows);
+
+    return remainingRowsFilter.filter(records);
+  }
+
+  private Filter<T> getFilter(Predicate<T> remainingRows) {
+    return new Filter<T>() {
       @Override
       protected boolean shouldKeep(T item) {
         return remainingRows.test(item);
       }
     };
+  }
 
-    return remainingRowsFilter.filter(records);
+  private Filter<T> getMarker(Predicate<T> remainingRows) {
+    return new Filter<T>() {
+      @Override
+      protected boolean shouldKeep(T item) {
+        if (!remainingRows.test(item)) {
+          // if the row is deleted, set the metadata column to true
+          markRowDeleted(item);
+        }
+        return true;  // keep the row even if it is deleted by equality deletions
+      }
+    };
+  }
+
+  protected void markRowDeleted(T item) {
```

Comment:
should this just be an abstract method and force impl classes to implement this method?

Reply:
Probably, it needs changes for all subclass of DeleteFilter. This is PR is already big. Maybe we should split it. For example, we can have a separated PR to read deleted rows in case of row-level change. cc @chenjunjiedada

Reply:
The previous PR which read the deleted rows has such implementation, +1 to use another PR.